File: LeetCode-Python/EvaluateDivision.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
         # 1) Grafı kur: her kenarın tersi de eklenir
        graph: Dict[str, List[Tuple[str, float]]] = {}
        for (a, b), v in zip(equations, values):
            graph.setdefault(a, []).append((b, v))
            graph.setdefault(b, []).append((a, 1.0 / v))

        def dfs(src: str, dst: str) -> float:
            # src veya dst grafikte yoksa
            if src not in graph or dst not in graph:
                return -1.0
            # aynı değişkense ve tanımlıysa
            if src == dst:
                return 1.0

            visited = set([src])
            stack: List[Tuple[str, float]] = [(src, 1.0)]  # (node, çarpan)

            while stack:
                node, acc = stack.pop()
                if node == dst:
                    return acc
                for nei, w in graph.get(node, []):
                    if nei in visited:
                        continue
                    visited.add(nei)
                    stack.append((nei, acc * w))
            return -1.0

        # 2) Sorguları çöz
        return [dfs(c, d) for c, d in queries]

File: LeetCode-Cpp/EvaluateDivision.cpp
class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        // Grafik: node -> [(komsu, oran)]
        unordered_map<string, vector<pair<string, double>>> g;
        g.reserve(equations.size() * 2);

        // 1) Grafı kur
        for (size_t i = 0; i < equations.size(); ++i) {
            const string& A = equations[i][0];
            const string& B = equations[i][1];
            double v = values[i];

            g[A].push_back({B, v});       // A / B = v
            g[B].push_back({A, 1.0 / v}); // B / A = 1/v
        }

        // 2) Her sorguyu çöz
        vector<double> ans;
        ans.reserve(queries.size());

        for (auto& q : queries) {
            const string& C = q[0];
            const string& D = q[1];

            // Değişkenlerden biri tanımsızsa
            if (!g.count(C) || !g.count(D)) {
                ans.push_back(-1.0);
                continue;
            }

            // Aynı değişkense ve grafikte varsa 1.0
            if (C == D) {
                ans.push_back(1.0);
                continue;
            }

            // DFS ile C'den D'ye yol ve çarpım ara
            unordered_set<string> vis;
            double result = -1.0;
            if (dfs(C, D, 1.0, vis, g, result)) {
                ans.push_back(result);
            } else {
                ans.push_back(-1.0);
            }
        }

        return ans;
    }
    private:
    bool dfs(const string& cur, const string& target, double acc,
             unordered_set<string>& vis,
             unordered_map<string, vector<pair<string, double>>>& g,
             double& out) {
        if (cur == target) {
            out = acc;
            return true;
        }
        vis.insert(cur);

        for (auto& [nxt, w] : g[cur]) {
            if (vis.count(nxt)) continue;
            if (dfs(nxt, target, acc * w, vis, g, out)) {
                return true;
            }
        }
        return false;
    }

};
